src/client/Microsoft.Identity.Client/Platforms/Features/WamBroker/MsaPassthroughHandler.cs

```diff
@@ -0,0 +1,146 @@
+        {
+            _logger.Verbose("WAM MSA-PT - fetching transfer token");
+            string transferToken = await FetchMsaPassthroughTransferTokenAsync(
+                authenticationRequestParameters, accountProvider)
+                .ConfigureAwait(false);
+
+            return transferToken;
+        }
+
+        public void AddTransferTokenToRequest(
+            Windows.Security.Authentication.Web.Core.WebTokenRequest webTokenRequest,
+            string transferToken)
+        {
+            if (!string.IsNullOrEmpty(transferToken))
+            {
+                webTokenRequest.Properties.Add("SamlAssertion", transferToken);
+                webTokenRequest.Properties.Add("SamlAssertionType", "SAMLV1");
+            }
+        }
+
+        private async Task<string> FetchMsaPassthroughTransferTokenAsync(
+           AuthenticationRequestParameters authenticationRequestParameters,
+           WebAccountProvider accountProvider)
+        {
+            // step 1 - get a response from the MSA provider, just to have a WebAccount
+            _logger.Info("WAM MSA-PT: Making initial call to MSA provider");
+            WebAccount msaPtWebAccount = await FetchWebAccountFromMsaAsync(
+                authenticationRequestParameters,
+                accountProvider).ConfigureAwait(false);
+
+            // step 2 - get a trasnfer token 
+            _logger.Info("WAM MSA-PT: Getting transfer token");
+            string transferToken = await FetchTransferTokenAsync(
+                accountProvider,
+                msaPtWebAccount,
+                authenticationRequestParameters.ClientId).ConfigureAwait(true);
+
+            return transferToken;
+        }
+
+        private async Task<WebAccount> FetchWebAccountFromMsaAsync(
+            AuthenticationRequestParameters authenticationRequestParameters, WebAccountProvider accountProvider)
+        {
+            // This response has an v1 MSA Access Token, which MSAL should expose to the user
+            var webTokenRequestMsa = await _msaPlugin.CreateWebTokenRequestAsync(
+                     accountProvider,
+                     authenticationRequestParameters,
+                     isForceLoginPrompt: false,
+                     isInteractive: true,
+                     isAccountInWam: false)
+                    .ConfigureAwait(false);
+
+            WamAdapters.AddMsalParamsToRequest(authenticationRequestParameters, webTokenRequestMsa);
+
+            var webTokenResponseMsa = await _wamProxy.RequestTokenForWindowAsync(_parentHandle, webTokenRequestMsa)
+                .ConfigureAwait(true);
+
+            if (!webTokenResponseMsa.ResponseStatus.IsSuccessStatus())
+            {
+                var errorResp = WamAdapters.CreateMsalResponseFromWamResponse(webTokenResponseMsa, _msaPlugin, _logger, true);
+                throw new MsalServiceException(
+                    errorResp.Error,
```

Comment:
>errorResp [](start = 20, length = 9)

can this be null?

Reply:
no, in that adapter object errorResp.Error is always assigned to. I hope :)

Reply:
i hope too.

---
In reply to: [565512297](https://github.com/AzureAD/microsoft-authentication-library-for-dotnet/pull/2365#discussion_r565512297) [](ancestors = 565512297)

Reply:
Isn't nullable great? I wish we could use it in MSAL.